### ISSUE TITLE
Functions to obtain ClientInfo

### ISSUE THREAD

**Issue by alexey-milovidov** [ClickHouse/ClickHouse#18408]
See `ClientInfo.h` and its usage in `system.query_log` (`QueryLog.h`).

**Use case**
When multiple clients are sending INSERT queries, we may want to have a materialized column with e.g. client hostname to automatically record where INSERT comes from.

**Describe the solution you'd like**
A single function `clientInfo('attr_name')` (its return type will depend on attr name) or a set of functions.

**Describe alternatives you've considered**
It's always doable with custom API middleware.

**Comment by spongedu** [comment]
@alexey-milovidov  Hi, I'd like to try on this if its ok. :)

**Comment by alexey-milovidov** [comment]
The function should be named `getClientInfo` to be similar to `getSetting`.

**Comment by MainHanzo** [comment]
@alexey-milovidov  Hello i would like to try to implement this feature since this pr has been inactive for a long time.
before starting it, i want to make sure I understand it correctly: 
1. the methid getClientInfo is supposed to be implement inside class ClientInfo right?
2. which getSetting method do you refer?

I am taking baby steps in this project, it would be great if you could provide more insight about this feature. Thank you very much. :)

**Comment by alexey-milovidov** [comment]
@MainHanzo thank you.

`getSetting` is an SQL function, and its implementation is located inside `src/Functions`
`getClientInfo` should also be an SQL function.

**Comment by MainHanzo** [comment]
@alexey-milovidov Hello alexey, I have one question concerning how we get the attribute value inside of ClientInfo class. I am not sure how can i reference queryLog.cpp in this case: How should I get the attribute value dynamically? I have checked the QueryLogElement::appendClientInfo method but what I see is only getting the client info attribut value statically. should I use the accessor(like what we do in BaseSettings.h)? Cus I can't find the implementation of 'accessor' so I am not sure if I am right. Thank you in advance for giving me more insight! :)

**Comment by Algunenano** [comment]
> . I am not sure how can i reference queryLog.cpp in this case: How should I get the attribute value dynamically? I have checked the QueryLogElement::appendClientInfo method but what I see is only getting the client info attribut value statically

You shouldn't be trying to access queryLog. The clientInfo of the running query is under Context (see `getContext()->getClientInfo()`).

**Comment by MainHanzo** [comment]
> > . I am not sure how can i reference queryLog.cpp in this case: How should I get the attribute value dynamically? I have checked the QueryLogElement::appendClientInfo method but what I see is only getting the client info attribut value statically
> 
> You shouldn't be trying to access queryLog. The clientInfo of the running query is under Context (see `getContext()->getClientInfo()`).

Yeah I agree. I was talking about how ClientInfo get its own attributes, more specifically after I get the ClientInfo instance by getContext()->getClientInfo(), how does the ClientInfo instance get the values of all the attributes it has

**Comment by Algunenano** [comment]
> how does the ClientInfo instance get the values of all the attributes it has

I don't understand what you mean. `getContext()->getClientInfo().client_hostname` will get you client_hostname for example, but I guess that's not what you are asking for. If what you want is an easy way to match any string to a member of the struct then I'd use an unordered map (string -> callback) where callback is a function to returns the stringify value, and that you'll need to define for each member (probably a one-line lambda would be best for most of them).

**Comment by MainHanzo** [comment]
> > how does the ClientInfo instance get the values of all the attributes it has
> 
> I don't understand what you mean. `getContext()->getClientInfo().client_hostname` will get you client_hostname for example, but I guess that's not what you are asking for. If what you want is an easy way to match any string to a member of the struct then I'd use an unordered map (string -> callback) where callback is a function to returns the stringify value, and that you'll need to define for each member (probably a one-line lambda would be best for most of them).

@Algunenano  THank you for your advice it was what I meant to ask.
I realized that we need to create a 'reflexion' of an object and to get all different types of member of clientInfo.h I need to do something similar as the solution in this blog: [https://www.codeproject.com/Articles/5359493/Design-Patterns-RunTime-Reflection-Cplusplus?msg=5969878#xx5969878xx]

I am not sure if this is a optimal solution but when i try to implement that, I was stuck a bug 'Redundant parentheses surrounding declarator' on the class memeber pointer.  I would be really grateful if some could give some help on this bug. Otherwise I think I might be the right direction and this solution seems to be too complicated.

**Comment by MainHanzo** [comment]
some complementary info about the bug I mentioned:
![1](https://github.com/ClickHouse/ClickHouse/assets/33153091/ec1cebdd-8e15-4b92-b9f6-c3c16f862ec6)

**Comment by alexey-milovidov** [comment]
We can use a blatantly simple switch/case block encapsulated inside a function.

**Comment by alexey-milovidov** [comment]
We didn't grow up to use reflection in C++...

**Comment by MainHanzo** [comment]
Thank you both for the guide. Another question which involves some more details: About the OpenTelemetry::TracingContext client_trace_context, when we query on this member by getClientInfo("client_trace_context"), what value should I return?

**Comment by MainHanzo** [comment]
> Thank you both for the guide. Another question which involves some more details: About the params like OpenTelemetry::TracingContext client_trace_context and Poco::Net::NameValueCollection headers, when we query on this member by getClientInfo("client_trace_context"), what value should I return?

@alexey-milovidov any more info on this? what if the attribute is not of simple type? what kind of info should we return?

**Comment by MainHanzo** [comment]
@Algunenano @alexey-milovidov Thanks for your advice. I am currently adding corresponding tests but i am little confused how I should test it.  According to https://clickhouse.com/docs/en/development/tests I think i need to add a .sh test but how should I add custom client info during the test? Any command that I might use?

**Comment by alexey-milovidov** [comment]
You can use curl to specify HTTP headers, such as User-Agent.

**Comment by MainHanzo** [comment]
@Algunenano @alexey-milovidov , hello and happy new year. Sorry for inactivity during the end of year. I am currently stuck by a problem of building my version with getClientInfo function. Brefly explaining what i am doing here: I implemented the getClientInfo function and i try to test it with a unit test, but when i build my code following this guide [https://clickhouse.com/docs/en/development/developer-instruction](url) i encountered some Cmake errors. I am building with macos clion. Any tips for building a new version locally? should I try with windows or other systems, is there any guide for building process? thx

**Comment by alexey-milovidov** [comment]
@MainHanzo, build works on Mac OS and Linux, and should be run precisely as in the instruction.

**Comment by alexey-milovidov** [comment]
This task is open, and we need it a lot!

**Comment by kiran-4444** [comment]
Hey, @MainHanzo are you still onn it? I'd like to take this up if you're not working on this.

**Comment by MainHanzo** [comment]
yes go ahead. I am not currently working on this

**Comment by sankalpsthakur** [comment]
I'd like to implement `getClientInfo` as discussed (SQL function analogous to `getSetting`, switch/case over ClientInfo scalar fields). Opening a PR shortly.
